internal/worker/apiaddresssetter/worker.go

```diff
@@ -0,0 +1,367 @@
+}
+
+// watchForControllerChanges starts a watcher for changes to controller nodes.
+// It returns a channel which will receive events if any of the watchers fires.
+func (w *apiAddressSetterWorker) watchForControllerNodeChanges() (<-chan struct{}, error) {
+	watcher, err := w.config.ControllerNodeService.WatchControllerNodes()
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+	if err := w.catacomb.Add(watcher); err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	return watcher.Changes(), nil
+}
+
+// watchForConfigChanges starts a watcher for changes to controller config.
+// It returns a channel which will receive events if the watcher fires.
+func (w *apiAddressSetterWorker) watchForConfigChanges() (<-chan []string, error) {
+	watcher, err := w.config.ControllerConfigService.WatchControllerConfig()
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+	if err := w.catacomb.Add(watcher); err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	return watcher.Changes(), nil
+}
+
+// updateControllerNodes updates the current list of tracked controller nodes,
+// as well as starting and stopping trackers for them as they are added and
+// removed.
+func (w *apiAddressSetterWorker) updateControllerNodes(ctx context.Context) (bool, error) {
+	controllerIDs, err := w.config.ControllerNodeService.GetControllerIDs(ctx)
+	if err != nil {
+		return false, errors.Errorf("cannot get controller IDs: %w", err)
+	}
+
+	w.config.Logger.Debugf(ctx, "controller nodes: %#v", controllerIDs)
+
+	var changed bool
+
+	// Stop controller tracker that no longer correspond to controller nodes.
+	workerNames := w.runner.WorkerNames()
+	for _, controllerID := range workerNames {
+		if !slices.Contains(controllerIDs, controllerID) {
+			if err := w.stopAndRemoveTracker(ctx, controllerID); err != nil {
+				return false, errors.Capture(err)
+			}
+			changed = true
+		}
+	}
+
+	// Start trackers for new nodes.
+	for _, controllerID := range controllerIDs {
+		w.config.Logger.Debugf(ctx, "found new controller %q", controllerID)
+
+		if err := w.runner.StartWorker(ctx, controllerID, func(ctx context.Context) (worker.Worker, error) {
+			id, err := strconv.Atoi(controllerID)
+			if err != nil {
+				return nil, errors.Errorf("invalid controller ID %q: %w", controllerID, err)
+			}
+			unitName, err := unit.NewNameFromParts(application.ControllerApplicationName, id)
+			if err != nil {
+				return nil, errors.Errorf("invalid unit name for controller %q: %w", controllerID, err)
+			}
+			tracker, err := newControllerTracker(unitName, w.config.ApplicationService, w.controllerNodeChanges, w.config.Logger.Child("controllertracker"))
+			if err != nil {
+				return nil, errors.Capture(err)
+			}
+			return tracker, nil
+		}); err != nil && !errors.Is(err, coreerrors.AlreadyExists) {
+			return false, errors.Errorf("failed to start tracker for controller node %q: %w", controllerID, err)
+		}
+		changed = true
```

Comment:
This is always going to be true, so I think we can drop this as it's tautological.

i.e. `return changed || len(controllerIDs)`